go/vt/dbconfigs/dbconfigs.go

```diff
@@ -24,191 +26,256 @@ import (
+func RegisterFlags(names ...string) {
+	registerBaseFlags()
+	for _, name := range names {
+		uc := &userConfig{}
+		dbConfigs.userConfigs[name] = uc
+		registerUserFlags(uc, name)
 	}
-	if AppDebugConfig&flags != 0 {
-		registerConnFlags(&dbConfigs.AppDebug, "appdebug")
-		registeredFlags |= AppDebugConfig
-	}
-	if AllPrivsConfig&flags != 0 {
-		registerConnFlags(&dbConfigs.AllPrivs, "allprivs")
-		registeredFlags |= AllPrivsConfig
-	}
-	if DbaConfig&flags != 0 {
-		registerConnFlags(&dbConfigs.Dba, "dba")
-		registeredFlags |= DbaConfig
-	}
-	if FilteredConfig&flags != 0 {
-		registerConnFlags(&dbConfigs.Filtered, "filtered")
-		registeredFlags |= FilteredConfig
-	}
-	if ReplConfig&flags != 0 {
-		registerConnFlags(&dbConfigs.Repl, "repl")
-		registeredFlags |= ReplConfig
-	}
-	return registeredFlags
 }
 
-// initConnParams may overwrite the socket file,
-// and refresh the password to check that works.
-func initConnParams(cp *mysql.ConnParams, socketFile string) error {
-	// Always try to connect with the socket if provided.
-	if socketFile != "" {
-		cp.UnixSocket = socketFile
-	}
+func registerBaseFlags() {
+	flag.StringVar(&baseConfig.UnixSocket, "db_socket", "", "The unix socket to connect on. If this is specifed, host and port will not be used.")
+	flag.StringVar(&baseConfig.Host, "db_host", "", "The host name for the tcp connection.")
+	flag.IntVar(&baseConfig.Port, "db_port", 0, "tcp port")
+	flag.StringVar(&baseConfig.Charset, "db_charset", "utf8", "Character set. Only utf8 or latin1 based character sets are supported.")
+	flag.Uint64Var(&baseConfig.Flags, "db_flags", 0, "Flag values as defined by MySQL.")
+	flag.StringVar(&baseConfig.SslCa, "db_ssl_ca", "", "connection ssl ca")
+	flag.StringVar(&baseConfig.SslCaPath, "db_ssl_ca_path", "", "connection ssl ca path")
+	flag.StringVar(&baseConfig.SslCert, "db_ssl_cert", "", "connection ssl certificate")
+	flag.StringVar(&baseConfig.SslKey, "db_ssl_key", "", "connection ssl key")
+}
 
-	// See if the CredentialsServer is working. We do not use the
-	// result for anything, this is just a check.
-	_, err := WithCredentials(cp)
-	return err
-}
-
-// DBConfigs is all we need for a smart tablet server:
-// - App access with db name for serving app queries
-// - AllPrivs access for administrative actions (like schema changes)
-//   that should be done without SUPER privilege
-// - Dba access for any dba-type operation (db creation, replication, ...)
-// - Filtered access for filtered replication
-// - Replication access to change master
-// - SidecarDBName for storing operational metadata
-type DBConfigs struct {
-	App           mysql.ConnParams
-	AppDebug      mysql.ConnParams
-	AllPrivs      mysql.ConnParams
-	Dba           mysql.ConnParams
-	Filtered      mysql.ConnParams
-	Repl          mysql.ConnParams
-	SidecarDBName string
+// The flags will change the global singleton
+// TODO(sougou): deprecate the legacy flags.
+func registerUserFlags(dbc *userConfig, name string) {
```

Comment:
+ Rename to registerPerUserFlags?
+ Rename "name" to "username"?

Reply:
I can call it 'key` or `userKey`, because the actual username is a field inside the struct.